zarr/indexing.py

```diff
@@ -0,0 +1,676 @@
+                           if not isinstance(s, IntDimIndexer))
+        self.is_advanced = any([not isinstance(dim_indexer, (IntDimIndexer, SliceDimIndexer))
+                                for dim_indexer in self.dim_indexers])
+        if self.is_advanced:
+            self.drop_axes = tuple([i for i, dim_indexer in enumerate(self.dim_indexers)
+                                       if isinstance(dim_indexer, IntDimIndexer)])
+        else:
+            self.drop_axes = None
+
+    def __iter__(self):
+        for dim_projections in itertools.product(*self.dim_indexers):
+
+            chunk_coords = tuple(p.dim_chunk_ix for p in dim_projections)
+            chunk_selection = tuple(p.dim_chunk_sel for p in dim_projections)
+            out_selection = tuple(p.dim_out_sel for p in dim_projections if p.dim_out_sel is not None)
+
+            # handle advanced indexing arrays orthogonally
+            if self.is_advanced:
+                # numpy doesn't support orthogonal indexing directly as yet, so need to work
+                # around via np.ix_. Also np.ix_ does not support a mixture of arrays and slices
+                # or integers, so need to convert slices and integers into ranges.
+                chunk_selection = ix_(*chunk_selection)
+
+                # special case for non-monotonic indices
+                if any([not isinstance(s, (int, slice)) for s in out_selection]):
+                    out_selection = ix_(*out_selection)
+
+            yield ChunkProjection(chunk_coords, chunk_selection, out_selection)
+
+
+class OIndex(object):
+
+    def __init__(self, array):
+        self.array = array
+
+    def __getitem__(self, selection):
+        return self.array.get_orthogonal_selection(selection)
+
+    def __setitem__(self, selection, value):
+        return self.array.set_orthogonal_selection(selection, value)
+
+
+def is_coordinate_selection(selection, array):
+    return (
+        (len(selection) == len(array._shape)) and
+        all(
+            [is_integer(dim_sel) or is_integer_array(dim_sel)
+             for dim_sel in selection]
+        )
+    )
+
+
+def is_mask_selection(selection, array):
+    return (
+        len(selection) == 1 and
+        is_bool_array(selection[0]) and
+        selection[0].shape == array.shape
+    )
+
+
+def replace_lists(selection):
+    return tuple(
+        np.asarray(dim_sel) if isinstance(dim_sel, list) else dim_sel
+        for dim_sel in selection
+    )
+
+
+# noinspection PyProtectedMember
+class CoordinateIndexer(object):
+
+    def __init__(self, selection, array):
+
+        # some initial normalization
+        selection = ensure_tuple(selection)
+        selection = tuple([i] if is_integer(i) else i for i in selection)
+        selection = replace_lists(selection)
+
+        # validation
+        if not is_coordinate_selection(selection, array):
+            # TODO refactor error messages for consistency
+            raise IndexError('invalid coordinate selection')
```

Comment:
It would be good to add an informative error message here about slices, because assuredly somebody is going to try that. (For what it's worth, I agree that it's a good choice not to support them!)

Reply:
Agreed, I'll do some work on error messages when implementation has settled.